keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,459 @@
+
+## Motivation
+
+- Enable `requiredDuringSchedulingRequiredDuringExecution` node affinity for pod scheduling and execution.
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+
+- Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **nodeAffinity** type.
+- Add `node-affinity-eviction` controller to ensure pods being evicted if the selectors are no longer met at runtime.
+- Explore a fine way to bring more descheduler features into core kubernetes(e.g. podAffinityEviction)
+  
+### Non-Goals
+
+- This KEP does not aim to introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAffinity** type and **podAntiAffinity** type.
+
+## Proposal
+- Implement the `requiredDuringSchedulingRequiredDuringExecution` node affinity type.
+- Update the kube-scheduler, kube-controller-manager and kube-apiserver.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories (Optional)
+
+#### manage multi tenant cluster with dedicated nodes
+
+I have some nodes labeled with "userA=allow" and "userB=allow", which means that these nodes are available to user A and user B.
+As the scale grows I want to add new nodes only for user B while leaving the current nodes only for user A.
+I want to migrate all pods available to user B to new nodes, and I want to ensure HA of my services during the migration.
+
+Without `node-affinity-eviction`, I have to remove "userB=allow" label of node, and delete the pods manually.
+With `node-affinity-eviction`, I can simply delete the "userB=allow" label from the existing nodes 
+to re-schedule all pods of user B to these new nodes.
+
+#### handle data locality pods when data is corrupted
+
+I have one application whose pods need to be co-located with some specific data stored in some particular nodes. 
+When one node's storage is corrupted, the pod on this node will provide some inaccurate services.
+So I want it to be scheduled to another special node when data is detected to be corrupted.
+And in the meantime, I want to keep other running pods on that node intact.
+Also, I want to ensure my PDB is guaranteed, so I can't taint the node directly.
+
+[stack overflow link](https://stackoverflow.com/questions/55006205/how-to-prevent-pods-scheduling-to-a-node-after-a-pod-has-failed-on-it-in-gce)
+
+Without `node-affinity-eviction`, I have to relabel the node to suggest data on that node is corrupted, 
+and then manually iterating the pods on the node, decide whether to delete them to trigger a reschedule.
+With `node-affinity-eviction`, the only thing I need to do is to relabel the node, and `node-affinity-eviction` controller will do the rest.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
```

Comment:
this should mention scenarios when the pod fails to be evicted and how it should be observed and resolved

Reply:
> how it should be observed

I haven't actually thought this part through yet.

So, at first my thought is to manually append an Condition for target pod.
And we decided to let Eviction API handles the condition. And I checked yesterday, seems the condition added by api will disappear after eviction failed? This doesn't seem to be easily observable by users.

I'm still thinking how we should improve its observability when the pod fails to be evicted.
Adding a metric with the label of pod's name?
Create a event to the cluster?

What's your opinion?

Reply:
- if the eviction fails the Condition does not appear
- a metric could help, not sure about the events